### PR TITLE
tgui-next: Initial PR, speedy new Air Alarm interface.

### PR DESCRIPTION
**Pull request by stylemistake**
## About The Pull Request

Read Issue #46246 

Some interfaces (one of them is Air Alarm UI) were rewritten in a completely new UI framework based on Inferno. New UI is very fast.

The rest of not implemented yet interfaces are delegated to an old tgui by dynamically loading old tgui js/css. Dynamic loading might slow down old tgui interfaces.

## Why It's Good For The Game

**[100% LAG FREE SCRUBBERS](https://cdn.discordapp.com/attachments/326831214667235328/626653924975443968/2019-09-26_08-37-48.mp4)**

## Changelog
:cl:
add: New React-based TGUI, which is INCREDIBLY FAST.
add: New Acclimator interface (@actioninja).
add: New AI Airlock interface (@actioninja).
add: New Air Alarm interface (@stylemistake).
add: New Airlock Electronics interface (@actioninja).
add: New APC interface (@actioninja).
add: New Atmos Alert Console interface (@actioninja).
add: New Atmos Control Console interface (@actioninja).
add: New Atmos Filter interface (@actioninja).
add: New Atmos Mixer interface (@actioninja).
add: New Atmos Pump interface (@actioninja).
add: New Borg Panel interface (@actioninja).
add: New Brig Timer interface (@actioninja).
add: New Bluespace Artillery interface (@actioninja).
add: New Gas Canister interface (@willox).
add: New Cargo interface (@actioninja, @stylemistake).
add: New Cellular Emporium interface for changelings (@actioninja).
add: New CentCom Pod Launcher interface (@actioninja).
add: New Chem Dispenser interface (@stylemistake).
add: New Crayon interface (@actioninja).
add: New Disposal Unit interface (@actioninja).
add: New Thermomachine interface (@willox).
add: New Wires (hacking) interface (@actioninja).
tweak: Old TGUI interfaces are dynamically loaded if new TGUI interface was not found. This might increase load times while we are undergoing the transition.
tweak: Scrubber list is more compact, using chemical formula notation for common gases.
tweak: Slightly more native-looking title bar in new TGUI.
tweak: Window scrollbars are now themed to match TGUI color scheme.
fix: New TGUI hopefully fixes all previous window dragging and resizing issues.
code: Very robust UI development pipeline, with hot module replacement and live log collection.
/:cl:

### CODE REVIEW

**Comment by Qustinnus** [comment]
I dislike the idea of introducing another UI system without removing the old, how will we prevent people from ignoring the old templates forever and then prevent those templates from causing maintenance issues?

**Comment by Qustinnus** [comment]
Besides that epic pr keep it up gamer

**Comment by stylemistake** [comment]
> I dislike the idea of introducing another UI system without removing the old, how will we prevent people from ignoring the old templates forever and then prevent those templates from causing maintenance issues?

If we don't keep both systems in place, we might just never transition to the new system, because the amount of work necessary to move all interfaces is pretty huge.

**Comment by Qustinnus** [comment]
okay but like we have 3 formats right now, I see it as a possible problem. I think straight out banning tgui changes is a good idea

File: tgui-next/packages/functional/index.js
Comment (JJRcop) [inline tgui-next/packages/functional/index.js]: Highlighting license for maintainers to review

**Comment by JJRcop** [comment]
Would the autodoc system work with these JS files?

**Comment by 81Denton** [comment]
@Qustinnus If this proves to be a viable full replacement, we can just change our CONTRIBUTING.md

**Comment by willox** [comment]
I don't know if it is intentional but this lacks the behaviour from TGUI where the buttons dim a little bit with UI_UPDATE (the orange eye in the top-left corner) and grey out fully when disabled (the red eye.)

You can reproduce both by just standing different distances from the air alarm.

**Comment by actioninja** [comment]
I'm not a big fan of requiring yarn on top of node, but playing around with this a bit the hotloader and live logs are incredible.

File: tgui-next/packages/functional/index.js
Comment (stylemistake) [reply tgui-next/packages/functional/index.js]: @JJRcop it is compatible with AGPL v3, because of the GPL-2.0-**or-later**. I can remove licensing too, implicitly upgrading it to AGPL v3, but affero's extra clauses are pointless for javascript.

**Comment by stylemistake** [comment]
> Would the autodoc system work with these JS files?

@JJRcop yes.

> I don't know if it is intentional but this lacks the behaviour from TGUI where the buttons dim a little bit with UI_UPDATE or UI_DISABLED (the orange/red eye in the top-left corner). You can reproduce both by just standing different distances from the air alarm.

Not implemented yet. I think I'll bring a next patch of fixes today, fixing this in the process too.

---

Also, a daily reminder...

## pls test on IE8 who has wine

Thanks, this is for you @Cyprex in particular, because you got it working, and I would waste a lot of time setting it up. Ping me if you need a pre-built DM binary.

**Comment by stylemistake** [comment]
I think now we have a full feature parity with old TGUI.

- Dimming was addressed.
- Added a ProgressBar component, with CSS animations.
- Added a Toast component, without CSS animations yet, but functional.
- Window resizing, on all left to bottom sides of a window.

Remaining fixes:

- FPS drop if you load complex interfaces and do a lot of rerenders. Same problem existed on an old TGUI.
- ???

[Progress demo 1](https://cdn.discordapp.com/attachments/326831214667235328/627474339234709514/2019-09-28_14-58-02.mp4)
[Progress demo 2](https://cdn.discordapp.com/attachments/326831214667235328/627476837458640896/2019-09-28_15-07-18.mp4)

@actioninja was busy porting other interfaces (they will land in a different PR):

image

File: tgui-next/packages/tgui/interfaces/AirAlarm.js
Comment (willox) [inline tgui-next/packages/tgui/interfaces/AirAlarm.js]: The localStatusText values here are in the wrong order.

**Comment by PKPenguin321** [comment]
>Would the autodoc system work with these JS files?

It seems to already have documentation in `tgui-next/README.md`, though I haven't checked to see how comprehensive it is. I would hope it contains basic development instructions that developers can use as a reference for making new UIs, which was notoriously difficult with tgui.

**Comment by stylemistake** [comment]
I added "next" to the folder so that when we are done, we kebab the old one, and this one becomes simply "tgui".

**Comment by SpaceManiac** [comment]
I have IE8 set up, will test when able

File: tgui-next/yarn.lock
Comment (spookydonut) [inline tgui-next/yarn.lock]: should this file be in .gitignore or what

Comment (stylemistake) [reply tgui-next/yarn.lock]: No, because it's necessary for the package manager to behave deterministically, aka it precisely locks all dependency versions so that you always get predictable installs on any system, and it's not really updating that much.

**Comment by Cyprex** [comment]
Throwing up netcat on port 30001 gave me an IE 7 useragent for byond webviews, standalone IE(iexplore.exe) in the wineprefix gives IE 8. "nc -l -p 30001" gave
>User-Agent: Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; Trident/4.0; byond_5.0; .NET CLR 2.0.50727; .NET CLR 3.0.30729; .NET CLR 3.5.30729; .NET4.0C; .NET4.0E)

for byond webviews. Standalone IE in the same wineprefix produces

>User-Agent: Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; byond_5.0; .NET CLR 2.0.50727; .NET CLR 3.0.30729; .NET CLR 3.5.30729; .NET4.0C; .NET4.0E)

The requests from byond were triggered by replacing tgui.bundle.js with the following three-line snippet:
>var t = new XMLHttpRequest();
>t.open("GET", 'http://127.0.0.1:30001', 0);
t>.send();

**Comment by SpaceManiac** [comment]
@Cyprex A particular meta tag needs to be near the start of `<head>` for BYOND's embedded webview to use the most recent rendering engine it can:

```html
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
````

This is true to get IE11 features on an IE11-supporting system too.

**Comment by stylemistake** [comment]
Alright, I got very good news. New UI is now officially usable on IE8.

Bad news: Old UI is not loading due to unknown issues. So far [this have been the main suspect](https://github.com/ractivejs/ractive/issues/2343), although it's only a guess, because just about everything in IE8 is a blackbox. Can't debug anything at all.

I've added 2 new polyfills (console stubs and CSS object model) in addition to 3 already existing, and heavily reworked the build pipeline to account for oddities of ActiveX™ JScript and Webpack/Babel abusing `Object.defineProperty` pretty heavily if not keeping them under control. Polyfills are now located externally in public folder as separate files, which are loaded into the browser through use of conditional IE tags.

> A particular meta tag needs to be near the start of <head> for BYOND's embedded webview to use the most recent rendering engine it can

It still reports its User Agent as IE 7, no matter how you position this tag, and it does not affect rendering at all. We tried to detect visual differences between compatibility and no compatibility, and we found none.

**Comment by optimumtact** [comment]
given the lack of movement in a long time on tgui im happy to approve this pr provisionally, even if I think the buildchain suffers from similar problems that might lead down a similar path of never getting any updates in.

**Comment by stylemistake** [comment]
> Old UI is not loading due to unknown issues.

Small update: Old UI is not loading for me even on public servers.

**Comment by stylemistake** [comment]
Alright folks, we got IE8 completely sorted.

> Old UI is not loading due to unknown issues.

This was fixed by creating a separate page `tgui-fallback.html`, which only has scripts necessary to run the old tgui. One problem is that `ref` will be lost in the process because this is a plain HTML file. Solution is extremely retarded, but it works nicely: we redirect the webview to `tgui-fallback.html?ref=${ref}`, and in a new page we reconstruct the template by doing this:

```js
var ref = location.search.replace('?ref=', '');
document.getElementById('data').setAttribute('data-ref', ref);
```

We do not need to reconstruct the state, because backend will send a fresh state anyway after calling `tgui:initialize`.

This also by proxy solves #46847, here is an image of it working:
image

For Linux folks: `tgui-dev-tools` now fully work on Linux, with live compilation, cache reloader and live log collection (via basic HTTP requests). It tries to search for your BYOND cache folder in `~/.wine` and `~/Games/byond` wine prefixes. So you can enjoy all benefits of development on Windows, and get that UI done and tested quicker.

On this note, I call for testmerging, and no more features will be added to this PR.

File: tgui-next/.editorconfig
Comment (MrStonedOne) [inline tgui-next/.editorconfig]: Please fix

Comment (stylemistake) [reply tgui-next/.editorconfig]: Fix what exactly

File: tgui-next/.eslintrc.yml
Comment (MrStonedOne) [inline tgui-next/.eslintrc.yml]: Please set to `1tbs`

File: tgui-next/.eslintrc.yml
Comment (MrStonedOne) [inline tgui-next/.eslintrc.yml]: Please change the second index from `2` to `tab`

**Comment by MrStonedOne** [comment]
Please read through https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md and ensure your linter is not enforcing different coding styles from the rest of the project (where applicable).

You will not be the only one modifying these JS files. You **MUST** make this accessible to somebody who has never worked in js, never worked with managed web frameworks, and has only coded in DM, as that describes an uncomfortably large percent of our contributor base.

The first step to doing that is ensuring they don't get burned by conflicting style enforcement, so tgui-next's linter and editor config should have settings designed to be as close as possible to the rest of the project's style guidelines.

OR!

Maybe add the linter in a seperate PR so this PR doesn't get bogged down with arguments over that.

File: tgui-next/.editorconfig
Comment (MrStonedOne) [reply tgui-next/.editorconfig]: This project uses tabs.

You are adding tgui-next to this project, so it too should use tabs.

File: tgui-next/yarn.lock
Comment (MrStonedOne) [reply tgui-next/yarn.lock]: The true answer here is that YARN is being stupid and abusing the `.lock` extension in a way that creates confusion. To anybody unfamiliar with YARN or other equally stupid package managers, .lock is normally an extension used for temp per-session purposes, and not for long term storage.

File: tgui-next/packages/functional/index.js
Comment (MrStonedOne) [reply tgui-next/packages/functional/index.js]: >but affero's extra clauses are pointless for javascript.

No they aren't! .min files are considered object files and AGPL forces anybody using them to provide the original version.

File: tgui-next/.eslintrc.yml
Comment (stylemistake) [reply tgui-next/.eslintrc.yml]: perhaps we can just simply relax some of these rules, including this one

File: tgui-next/.eslintrc.yml
Comment (stylemistake) [reply tgui-next/.eslintrc.yml]: If i am not wrong, tools like npm and yarn use spaces regardless of tab usage in file or editorconfig setting, and spaces-as-tabs are used almost in every js project out there. If you're using snippets of javascript from the web, you will be forced to convert it to tabs. Old tgui also uses spaces instead of tabs, that's why I tried to reflect this in editor settings.

Linter is also capable of fixing styling issues automatically by running it with an extra `--fix` flag.

**Comment by stylemistake** [comment]
We can replace Yarn with Lerna:

https://github.com/lerna/lerna

This will allow us to continue to use a multi-package setup. This is still a separate tool, but it can be safely installed locally with npm, and we can add a post-install hook to bootstrap all packages. From a developer's perspective, you only have to write `npm install`.

Installation procedure will take a bit longer than yarn, though, and package-lock.json is still badly implemented in NPM, which might cause long term issues with dependencies randomly updating, etc.

**Comment by stylemistake** [comment]
I have rebased the branch, and replaced Yarn with Lerna. So essentially, you no longer need to install Yarn.

**Comment by optimumtact** [comment]
they should just learn to code

**Comment by stylemistake** [comment]
Testmerge is not rolling out well, because a lot of people still get an old tgui. All javascript and css assets are properly updated in the cache, the problem is that [an old HTML template](https://github.com/stylemistake/tgstation/blob/tgui-next/code/controllers/subsystem/tgui.dm#L13-L14) is served, which is a part of an rsc file.

This might be somehow related to this: http://www.byond.com/forum/post/103188

**Comment by optimumtact** [comment]
@MrStonedOne is your CDN fuxxored?

File: tgui-next/.eslintrc.yml
Comment (spookydonut) [reply tgui-next/.eslintrc.yml]: There are some files in this repo still using space indentation and it's shit.

**Comment by MrStonedOne** [comment]
The CDN only has resources, not assets, assets are handled by the asset cache and byond's `browse_rsc` system, and that doesn't cache byond the session. To be exact, the asset cache framework will cache per-connection, and DS will cache per-process-session, but caches based off of BYONDcrc, not filename.

It shouldn't be that, i might try to reproduce if somebody can give me a step to reproduce and check, mainly like what interface to open and how to know old vs new layout.

**Comment by stylemistake** [comment]
@MrStonedOne
> if somebody can give me a step to reproduce
There is no clear way to reproduce it, some of us got a new UI, some of us didn't. So far it's a total mistery.

> ...and check, mainly like what interface to open and how to know old vs new layout.
New Air Alarm UI can be easily identified by a glowing red close button when you hover it, as on [this screenshot](https://user-images.githubusercontent.com/1516236/65765905-1ea11b80-e132-11e9-8d13-2514c8a0c118.png).

**Comment by Cyprex** [comment]
Tested most of the UIs migrated recently on a local server. Everything seems to work in IE 8.

Testmerge has been updated on Terry by oranges.

**Comment by stylemistake** [comment]
@MrStonedOne nothing from the code side of things is helping with this cached html issue. I tried loading html in runtime, even deleted the old html file, and I still receive an old tgui on Terry (which is currently on commit 1d8859a). I really think it has something to do with deploy code. Please help me fix this because this is critically important and we can't merge it in this state.

**Comment by optimumtact** [comment]
@MrStonedOne there definiteyl seems to be some issue where they're being served the old html and cleaning up the entire local cache isn't helping, it's either an issue with tgs not doing a clean compile, or the cloudflare cdn is caching the base html and serving a different thing to different players.

**Comment by JJRcop** [comment]
How is the documentation as of now, does it adequately cover important functions and syntax of tgui2? How well are the internals documented for future contributors?

**Comment by stylemistake** [comment]
README has a pretty good documentation of UI components and what you can do with them.

Almost all helper functions in the "common" package have docstrings, and they are extensively used througout the code, so you learn about them by example.

I do admit that state management and Redux thing needs a little bit of explanation, because it's not very obvious. You can make some **really advanced** pieces of functionality with it. What we currently have is support for hotkeys and toast notifications. This is all largely undocumented and a work in progress.

If you a complete newbie however, you can ignore that it even exists. For all you care, it just provides your components with the state from the server (you get it via the `state` prop), and that's about it.

As of the internals, the build pipeline, I don't know if it's really that important to document. I mean, code is self explanatory for the most part (if you are willing to read it), there is very little magic happening behind the scenes.

**Comment by stylemistake** [comment]
> syntax of tgui2

README refers to React documentation, which explains pretty much everything, and if you are new to React, you are **required** to read it.

**Comment by stylemistake** [comment]
Full feature parity has been reached, IE8 works, additional failsafe was added to keep old tgui working.

Ready for the merge.

**Comment by Anonmare** [comment]
I don't think this is respecting APC modes as if the pwoer is low and I try to force equipment to ON, the APC will force it to OFF

**Comment by Cyprex** [comment]
d89f9de broke IE 8 support: Exception thrown and not caught (http://127.0.0.1:51789/tmp349/tgui.bundle.js line 1)

**Comment by actioninja** [comment]
It's most likely not due to any of my changes, but commits were squashes a ton anyways. That commit also isn't entirely my code.

**Comment by Cyprex** [comment]
That commit came up during bisection, could you link me to some source commits that I could check to further narrow down the source?

**Comment by actioninja** [comment]
you could try pulling `tgui-next-ui-dev` off of my branch and see if it's still broken. If it isn't, then it's more likely to be tweaks made by style himself.

**Comment by Cyprex** [comment]
fef0d06bbca45a6bbde9b26b1c4691d3833f8629 , "ui dev work" on that branch seems to have introducted this issue.

File: tgui-next/packages/tgui/interfaces/CentcomPodlauncher.js
Comment (TheChosenEvilOne) [inline tgui-next/packages/tgui/interfaces/CentcomPodlauncher.js]: hmmm, this does not look right.

Comment (stylemistake) [reply tgui-next/packages/tgui/interfaces/CentcomPodlauncher.js]: @actioninja

**Comment by JJRcop** [comment]
If maintainers believe the README needs more they will let you know. I've removed the label since you've created quite the amount of documentation, and now it's up to review.

**Comment by stylemistake** [comment]
Some nice-to-haves to implement later:
- Warn about unused `onClick` handler on buttons, because that is a potential bug, but I don't know how to set it up in such a way that would be annoying for devs when prototyping. Perhaps emit the warning on production builds, to be logged in round logs.
- Warn about lowercase `onclick` handlers, which is also a pretty common source of bugs.
- Emit errors when inserting illegal children elements, like `<br />` into `<LabeledList>` or inserting something else than a `<Tabs.Tab>` into `<Tabs>`.

File: tgui-next/packages/tgui/interfaces/AtmosControlConsole.js
Comment (willox) [inline tgui-next/packages/tgui/interfaces/AtmosControlConsole.js]: should be `onClick={() => act(ref, 'rate')} />`

File: code/modules/client/asset_cache.dm
Comment (AnturK) [inline code/modules/client/asset_cache.dm]: should use the new helper here for consistency.

**Review by AnturK — CHANGES_REQUESTED** [review]
DM side stuff, i will be looking over js shortly.

**Comment by Kraseo** [comment]
alien multitools do not show what wire does which when you hold it in hand with this interface, would be neat if it could be fixed

File: tgui-next/packages/tgui-dev-server/link/client.js
Comment (stylemistake) [reply tgui-next/packages/tgui-dev-server/link/client.js]: This code was used for debugging hot module replacement.

File: code/modules/client/asset_cache.dm
Comment (stylemistake) [reply code/modules/client/asset_cache.dm]: You mean we can't leave `icon_class_name` helper to be used for future interfaces? Just saying, Inferno can't render html tags like that.

**Comment by Cyprex** [comment]
I think borg wires are affected as well, I didn't get a legend with blueprints in the offhand

**Pull request merged**
